### ISSUE TITLE
No playback after redirects (MPD & HLS)

### ISSUE THREAD

**Issue by matthuisman** [peak3d/inputstream.adaptive#473]
This is not a new issue or regression. As far as I know, they have never worked.

However, dai.google.com urls are starting to redirect to urls for including ads etc, and that breaks playback if those mpds have baseurls

***HLS***

**Test without redirect (WORKS)**
https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_redirected/hls_direct.strm

**Test with redirect (DOES NOT WORK)**
https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_redirected/hls_redirected.strm

IA tries to get
`https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_redirected/matthuisman/ia_tests/master/hls_aes128_single_key/index_17_2744763.ts`

instead of correct
`https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_aes128_single_key/index_17_2744763.ts`

The test playlist used above
https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_redirected/master.m3u8
https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_redirected/sub.m3u8

Note:
Redirects with segments that don't start with / seem to work ok.

---

***MPD***

**Test without redirect (WORKS)**
https://raw.githubusercontent.com/matthuisman/ia_tests/master/mpd_redirected/mpd_direct.strm

**Test with redirect (NOT WORK)**
https://raw.githubusercontent.com/matthuisman/ia_tests/master/mpd_redirected/mpd_redirected.strm

IA tries to get:
`https://raw.githubusercontent.com/matthuisman/ia_tests/master/mpd_redirected/audio/mp4a/2second/tears_of_steel_1080p_audio_32k_dash_track1_init.mp4`

Instead of correct:
`https://dash.akamaized.net/dash264/TestCasesIOP33/Content/audio/mp4a/2second/tears_of_steel_1080p_audio_32k_dash_track1_init.mp4`

The test mpd used above
https://raw.githubusercontent.com/matthuisman/ia_tests/master/mpd_redirected/mpd_redirected.mpd

Note:
Redirects to a MPD without a BaseURL seem to work fine.

**Comment by smf007** [comment]
Regarding the HLS issue, seems to be related to the variable "base_domain".  That variable is populated in routine AdaptiveTree::PreparePaths.  When I get a chance this week, I'll add some debugging to see what's going on.

**Comment by matthuisman** [comment]
yeh, i looked into it.
IA tries to remember the original url and the redirected url.

It just needs to be cleaned up to on the 1st initial request for the url, if its redirected - then use the new url as the main url.
don't try to track both urls.  Dont need to keep the original url after the redirect.

**Comment by smf007** [comment]
@matthuisman is there a simple way to send a test playlist to IA?  Using the CBC addon with additional debug logging, the processing looks like it should work for the HLS issue.  If there was a quick way I could send your playlists that you linked above?

EDIT::  Basically I can only confirm the without redirect.  I don't have the understanding yet how the redirect is dealt with.

EDIT2:: As they are .strm files, just place a local copy and use the file-browser.

**Comment by smf007** [comment]
`DEBUG: CCurlFile::Open - effective URL: <https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_redirected/master.m3u8>`
`DEBUG: AddOnLog: InputStream Adaptive: Effective URL https://raw.githubusercontent.com/matthuisman/ia_tests/master/hls_redirected/master.m3u8`
`ERROR: AddOnLog: InputStream Adaptive: Source_URL: https://bit.ly/sub.m3u8`
`ERROR: AddOnLog: InputStream Adaptive: Base_URL: https://bit.ly/`
`ERROR: AddOnLog: InputStream Adaptive: Base_domain: https://bit.ly`
`ERROR: AddOnLog: InputStream Adaptive: mpdFileURL (https://bit.ly/2y4maMI)`

Basically, effective_url should be used instead of url.  This is already covered for some entries, but not all.  Looks like it needs to be added to all entries.

**Comment by matthuisman** [comment]
yes, on the initial request to get the master player, it needs to take the effective url (the actual final url) and then use that instead of source url. 

I see no reason why it trys to keep track of the original and effective. It should always just use the effective only. 

Once you do that, you can get rid of all the stuff it does later on where it checks if there is effective url, then try's to add that base domain to the path to the next file. If we use the effective url from the start, no need for all that.

Once its redirected, we don't need to know about the original url.

Relative urls are always based off the effective (final) url.
So, the only time the original is needed again, is the bit of logging it does in kodi.
However, it would be fine to log the redirected url of the playlist instead of the original requested on.

**Comment by smf007** [comment]
Pull #490 addresses the redirect for HLS including the URI key.  However this doesn't seem to carry over to the MPD.

**Comment by smf007** [comment]
Pull #491 addresses the redirect for MPD.

**Comment by matthuisman** [comment]
@smf007 

I'm working on something here:
https://github.com/matthuisman/inputstream.adaptive/commit/21480d5c2660f7be19e65cb5a1a9d8aed81e2133

Hopefully that makes sense :)
Makes it all a bit simpler

**Comment by matthuisman** [comment]
fixed with pr: https://github.com/peak3d/inputstream.adaptive/pull/508
